File: packages/python-sdk/plinto/webhooks.py
"""
Webhook management module for the Plinto SDK
"""

from typing import Dict, List, Optional, Any

from .http_client import HTTPClient
from .types import (
    WebhookEndpoint,
    WebhookEndpointCreateRequest,
    WebhookEndpointUpdateRequest,
    WebhookEvent,
    WebhookDelivery,
    WebhookEventType,
    WebhookEndpointListResponse,
)
from .utils import build_query_params, validate_webhook_signature


class WebhooksModule:
    """Webhook management operations"""
    
    def __init__(self, http_client: HTTPClient):
        self.http = http_client
    
    async def create_endpoint(
        self,
        request: WebhookEndpointCreateRequest
    ) -> WebhookEndpoint:
        """
        Create a new webhook endpoint
        
        Args:
            request: Webhook endpoint creation request
            
        Returns:
            Created webhook endpoint
        """
        response = await self.http.post(
            "/api/v1/webhooks/",
            json_data=request.dict()
        )
        return WebhookEndpoint(**response.json())
    
    async def list_endpoints(
        self,
        is_active: Optional[bool] = None
    ) -> WebhookEndpointListResponse:
        """
        List webhook endpoints for current user
        
        Args:
            is_active: Filter by active status
            
        Returns:
            List of webhook endpoints
        """
        params = build_query_params({"is_active": is_active})
        
        response = await self.http.get("/api/v1/webhooks/", params=params)
        data = response.json()
        
        return WebhookEndpointListResponse(
            endpoints=[WebhookEndpoint(**ep) for ep in data["endpoints"]],
            meta={
                "page": 1,
                "per_page": data["total"],
                "total": data["total"],
                "total_pages": 1
            }
        )
    
    async def get_endpoint(self, endpoint_id: str) -> WebhookEndpoint:
        """
        Get webhook endpoint details
        
        Args:
            endpoint_id: Webhook endpoint ID
            
        Returns:
            Webhook endpoint data
        """
        response = await self.http.get(f"/api/v1/webhooks/{endpoint_id}")
        return WebhookEndpoint(**response.json())
    
    async def update_endpoint(
        self,
        endpoint_id: str,
        request: WebhookEndpointUpdateRequest
    ) -> WebhookEndpoint:
        """
        Update webhook endpoint configuration
        
        Args:
            endpoint_id: Webhook endpoint ID
            request: Update request data
            
        Returns:
            Updated webhook endpoint
        """
        response = await self.http.patch(
            f"/api/v1/webhooks/{endpoint_id}",
            json_data=request.dict(exclude_unset=True)
        )
        return WebhookEndpoint(**response.json())
    
    async def delete_endpoint(self, endpoint_id: str) -> Dict[str, str]:
        """
        Delete webhook endpoint
        
        Args:
            endpoint_id: Webhook endpoint ID
            
        Returns:
            Success message
        """
        response = await self.http.delete(f"/api/v1/webhooks/{endpoint_id}")
        return response.json()
    
    async def test_endpoint(self, endpoint_id: str) -> Dict[str, str]:
        """
        Send test webhook to endpoint
        
        Args:
            endpoint_id: Webhook endpoint ID
            
        Returns:
            Success message
        """
        response = await self.http.post(f"/api/v1/webhooks/{endpoint_id}/test")
        return response.json()
    
    async def get_endpoint_stats(
        self,
        endpoint_id: str,
        days: int = 7
    ) -> Dict[str, Any]:
        """
        Get webhook endpoint delivery statistics
        
        Args:
            endpoint_id: Webhook endpoint ID
            days: Number of days to include in stats
            
        Returns:
            Delivery statistics
        """
        params = build_query_params({"days": days})
        
        response = await self.http.get(
            f"/api/v1/webhooks/{endpoint_id}/stats",
            params=params
        )
        return response.json()
    
    async def list_events(
        self,
        endpoint_id: str,
        limit: int = 100,
        offset: int = 0
    ) -> Dict[str, Any]:
        """
        List webhook events for an endpoint
        
        Args:
            endpoint_id: Webhook endpoint ID
            limit: Maximum events to return (1-1000)
            offset: Offset for pagination
            
        Returns:
            List of webhook events
        """
        params = build_query_params({
            "limit": limit,
            "offset": offset
        })
        
        response = await self.http.get(
            f"/api/v1/webhooks/{endpoint_id}/events",
            params=params
        )
        data = response.json()
        
        return {
            "events": [WebhookEvent(**event) for event in data["events"]],
            "total": data["total"]
        }
    
    async def list_deliveries(
        self,
        endpoint_id: str,
        limit: int = 100,
        offset: int = 0
    ) -> List[WebhookDelivery]:
        """
        List webhook delivery attempts for an endpoint
        
        Args:
            endpoint_id: Webhook endpoint ID
            limit: Maximum deliveries to return (1-1000)
            offset: Offset for pagination
            
        Returns:
            List of webhook deliveries
        """
        params = build_query_params({
            "limit": limit,
            "offset": offset
        })
        
        response = await self.http.get(
            f"/api/v1/webhooks/{endpoint_id}/deliveries",
            params=params
        )
        return [WebhookDelivery(**delivery) for delivery in response.json()]
    
    async def regenerate_secret(self, endpoint_id: str) -> WebhookEndpoint:
        """
        Regenerate webhook endpoint secret
        
        Args:
            endpoint_id: Webhook endpoint ID
            
        Returns:
            Updated webhook endpoint with new secret
        """
        response = await self.http.post(
            f"/api/v1/webhooks/{endpoint_id}/regenerate-secret"
        )
        return WebhookEndpoint(**response.json())
    
    async def get_available_event_types(self) -> List[WebhookEventType]:
        """
        List all available webhook event types
        
        Returns:
            List of available event types
        """
        response = await self.http.get("/api/v1/webhooks/events/types")
        return [WebhookEventType(event_type) for event_type in response.json()]
    
    async def verify_signature(
        self,
        secret: str,
        payload: str,
        signature: str
    ) -> Dict[str, bool]:
        """
        Verify webhook signature for testing
        
        Args:
            secret: Webhook endpoint secret
            payload: Raw webhook payload
            signature: Signature to verify
            
        Returns:
            Verification result
        """
        response = await self.http.post(
            "/api/v1/webhooks/verify-signature",
            json_data={
                "secret": secret,
                "payload": payload,
                "signature": signature
            }
        )
        return response.json()
    
    @staticmethod
    def validate_webhook_signature(
        payload: str,
        signature: str,
        secret: str,
        timestamp: Optional[str] = None,
        tolerance: int = 300
    ) -> bool:
        """
        Validate webhook signature locally
        
        Args:
            payload: Raw webhook payload string
            signature: Signature from X-Plinto-Signature header
            secret: Webhook endpoint secret
            timestamp: Timestamp from X-Plinto-Timestamp header
            tolerance: Maximum age of webhook in seconds
            
        Returns:
            True if signature is valid, False otherwise
        """
        return validate_webhook_signature(payload, signature, secret, timestamp, tolerance)

File: packages/typescript-sdk/src/webhooks.ts
/**
 * Webhook management module for the Plinto TypeScript SDK
 */

import type { HttpClient } from './http-client';
import type {
  WebhookEndpoint,
  WebhookEndpointCreateRequest,
  WebhookEndpointUpdateRequest,
  WebhookEvent,
  WebhookDelivery,
  WebhookEventType
} from './types';
import { ValidationError, WebhookError } from './errors';
import { ValidationUtils, WebhookUtils } from './utils';

/**
 * Webhook management operations
 */
export class Webhooks {
  constructor(private http: HttpClient) {}

  /**
   * Create a new webhook endpoint
   */
  async createEndpoint(request: WebhookEndpointCreateRequest): Promise<WebhookEndpoint> {
    // Validate input
    if (!ValidationUtils.isValidUrl(request.url)) {
      throw new ValidationError('Invalid webhook URL format');
    }

    if (!request.events || request.events.length === 0) {
      throw new ValidationError('At least one event type must be specified');
    }

    // Validate event types
    const validEvents = Object.values(WebhookEventType);
    const invalidEvents = request.events.filter(event => !validEvents.includes(event));
    if (invalidEvents.length > 0) {
      throw new ValidationError(`Invalid event types: ${invalidEvents.join(', ')}`);
    }

    if (request.description && request.description.length > 500) {
      throw new ValidationError('Description must be 500 characters or less');
    }

    // Validate custom headers
    if (request.headers) {
      const invalidHeaders = Object.keys(request.headers).filter(key => 
        key.toLowerCase().startsWith('authorization') || 
        key.toLowerCase().startsWith('x-webhook')
      );
      if (invalidHeaders.length > 0) {
        throw new ValidationError(`Reserved header names cannot be used: ${invalidHeaders.join(', ')}`);
      }
    }

    const response = await this.http.post<WebhookEndpoint>('/api/v1/webhooks/', request);
    return response.data;
  }

  /**
   * List webhook endpoints for current user
   */
  async listEndpoints(isActive?: boolean): Promise<{
    endpoints: WebhookEndpoint[];
    total: number;
  }> {
    const params: Record<string, any> = {};
    if (isActive !== undefined) {
      params.is_active = isActive;
    }

    const response = await this.http.get<{
      endpoints: WebhookEndpoint[];
      total: number;
    }>('/api/v1/webhooks/', { params });
    return response.data;
  }

  /**
   * Get webhook endpoint details
   */
  async getEndpoint(endpointId: string): Promise<WebhookEndpoint> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    const response = await this.http.get<WebhookEndpoint>(`/api/v1/webhooks/${endpointId}`);
    return response.data;
  }

  /**
   * Update webhook endpoint
   */
  async updateEndpoint(
    endpointId: string,
    request: WebhookEndpointUpdateRequest
  ): Promise<WebhookEndpoint> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    // Validate input
    if (request.url && !ValidationUtils.isValidUrl(request.url)) {
      throw new ValidationError('Invalid webhook URL format');
    }

    if (request.events) {
      if (request.events.length === 0) {
        throw new ValidationError('At least one event type must be specified');
      }

      const validEvents = Object.values(WebhookEventType);
      const invalidEvents = request.events.filter(event => !validEvents.includes(event));
      if (invalidEvents.length > 0) {
        throw new ValidationError(`Invalid event types: ${invalidEvents.join(', ')}`);
      }
    }

    if (request.description && request.description.length > 500) {
      throw new ValidationError('Description must be 500 characters or less');
    }

    if (request.headers) {
      const invalidHeaders = Object.keys(request.headers).filter(key => 
        key.toLowerCase().startsWith('authorization') || 
        key.toLowerCase().startsWith('x-webhook')
      );
      if (invalidHeaders.length > 0) {
        throw new ValidationError(`Reserved header names cannot be used: ${invalidHeaders.join(', ')}`);
      }
    }

    const response = await this.http.patch<WebhookEndpoint>(`/api/v1/webhooks/${endpointId}`, request);
    return response.data;
  }

  /**
   * Delete webhook endpoint
   */
  async deleteEndpoint(endpointId: string): Promise<{ message: string }> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    const response = await this.http.delete<{ message: string }>(`/api/v1/webhooks/${endpointId}`);
    return response.data;
  }

  /**
   * Test webhook endpoint
   */
  async testEndpoint(endpointId: string): Promise<{ message: string }> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    const response = await this.http.post<{ message: string }>(`/api/v1/webhooks/${endpointId}/test`);
    return response.data;
  }

  /**
   * Get webhook endpoint statistics
   */
  async getEndpointStats(endpointId: string, days = 7): Promise<{
    total_deliveries: number;
    successful: number;
    failed: number;
    success_rate: number;
    average_delivery_time: number;
    period_days: number;
  }> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    if (days < 1 || days > 90) {
      throw new ValidationError('Days must be between 1 and 90');
    }

    const response = await this.http.get(`/api/v1/webhooks/${endpointId}/stats`, {
      params: { days }
    });
    return response.data;
  }

  /**
   * List webhook events for an endpoint
   */
  async listEvents(
    endpointId: string,
    options?: {
      limit?: number;
      offset?: number;
    }
  ): Promise<{
    events: WebhookEvent[];
    total: number;
  }> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    const params: Record<string, any> = {};
    if (options?.limit !== undefined) {
      if (options.limit < 1 || options.limit > 1000) {
        throw new ValidationError('Limit must be between 1 and 1000');
      }
      params.limit = options.limit;
    }

    if (options?.offset !== undefined) {
      if (options.offset < 0) {
        throw new ValidationError('Offset must be non-negative');
      }
      params.offset = options.offset;
    }

    const response = await this.http.get<{
      events: WebhookEvent[];
      total: number;
    }>(`/api/v1/webhooks/${endpointId}/events`, { params });
    return response.data;
  }

  /**
   * List webhook delivery attempts for an endpoint
   */
  async listDeliveries(
    endpointId: string,
    options?: {
      limit?: number;
      offset?: number;
    }
  ): Promise<WebhookDelivery[]> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    const params: Record<string, any> = {};
    if (options?.limit !== undefined) {
      if (options.limit < 1 || options.limit > 1000) {
        throw new ValidationError('Limit must be between 1 and 1000');
      }
      params.limit = options.limit;
    }

    if (options?.offset !== undefined) {
      if (options.offset < 0) {
        throw new ValidationError('Offset must be non-negative');
      }
      params.offset = options.offset;
    }

    const response = await this.http.get<WebhookDelivery[]>(`/api/v1/webhooks/${endpointId}/deliveries`, {
      params
    });
    return response.data;
  }

  /**
   * Regenerate webhook endpoint secret
   */
  async regenerateSecret(endpointId: string): Promise<WebhookEndpoint> {
    if (!ValidationUtils.isValidUUID(endpointId)) {
      throw new ValidationError('Invalid endpoint ID format');
    }

    const response = await this.http.post<WebhookEndpoint>(`/api/v1/webhooks/${endpointId}/regenerate-secret`);
    return response.data;
  }

  /**
   * Get available webhook event types
   */
  async getEventTypes(): Promise<string[]> {
    const response = await this.http.get<string[]>('/api/v1/webhooks/events/types');
    return response.data;
  }

  /**
   * Verify webhook signature (for testing)
   */
  async verifySignature(
    secret: string,
    payload: string,
    signature: string
  ): Promise<{ valid: boolean }> {
    if (!secret || !payload || !signature) {
      throw new ValidationError('Secret, payload, and signature are all required');
    }

    const response = await this.http.post<{ valid: boolean }>('/api/v1/webhooks/verify-signature', {
      secret,
      payload,
      signature
    });
    return response.data;
  }

  // Client-side utilities

  /**
   * Verify webhook signature (client-side)
   */
  async verifyWebhookSignature(
    payload: string,
    signature: string,
    secret: string
  ): Promise<boolean> {
    try {
      return await WebhookUtils.verifySignature(payload, signature, secret);
    } catch (error) {
      throw new WebhookError('Failed to verify webhook signature', { originalError: error });
    }
  }

  /**
   * Parse webhook payload
   */
  parseWebhookPayload<T = any>(payload: string): {
    event: WebhookEventType;
    data: T;
    timestamp: string;
    id: string;
  } {
    try {
      const parsed = JSON.parse(payload);
      
      if (!parsed.event || !parsed.data || !parsed.timestamp || !parsed.id) {
        throw new WebhookError('Invalid webhook payload format');
      }

      if (!Object.values(WebhookEventType).includes(parsed.event)) {
        throw new WebhookError(`Unknown webhook event type: ${parsed.event}`);
      }

      return parsed;
    } catch (error) {
      if (error instanceof WebhookError) {
        throw error;
      }
      throw new WebhookError('Failed to parse webhook payload', { originalError: error });
    }
  }

  /**
   * Validate webhook endpoint URL
   */
  validateEndpointUrl(url: string): { valid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (!url) {
      errors.push('URL is required');
      return { valid: false, errors };
    }

    if (!ValidationUtils.isValidUrl(url)) {
      errors.push('Invalid URL format');
      return { valid: false, errors };
    }

    try {
      const urlObj = new URL(url);
      
      // Must use HTTPS in production
      if (urlObj.protocol !== 'https:' && urlObj.protocol !== 'http:') {
        errors.push('URL must use HTTP or HTTPS protocol');
      }

      // Recommend HTTPS
      if (urlObj.protocol === 'http:') {
        errors.push('HTTPS is recommended for webhook endpoints');
      }

      // Check for localhost/private IPs in production
      const hostname = urlObj.hostname;
      if (hostname === 'localhost' || hostname === '127.0.0.1' || hostname.endsWith('.local')) {
        errors.push('Localhost URLs are not recommended for production webhooks');
      }

      // Check for standard webhook path
      if (!urlObj.pathname.includes('webhook')) {
        errors.push('Consider using a path that includes "webhook" for clarity');
      }

    } catch (error) {
      errors.push('Invalid URL format');
    }

    return {
      valid: errors.length === 0,
      errors
    };
  }

  /**
   * Get webhook event categories
   */
  getEventCategories(): Record<string, WebhookEventType[]> {
    return {
      user: [
        WebhookEventType.USER_CREATED,
        WebhookEventType.USER_UPDATED,
        WebhookEventType.USER_DELETED,
        WebhookEventType.USER_SIGNED_IN,
        WebhookEventType.USER_SIGNED_OUT
      ],
      session: [
        WebhookEventType.SESSION_CREATED,
        WebhookEventType.SESSION_EXPIRED
      ],
      organization: [
        WebhookEventType.ORGANIZATION_CREATED,
        WebhookEventType.ORGANIZATION_UPDATED,
        WebhookEventType.ORGANIZATION_DELETED,
        WebhookEventType.ORGANIZATION_MEMBER_ADDED,
        WebhookEventType.ORGANIZATION_MEMBER_REMOVED
      ]
    };
  }

  /**
   * Get friendly event type names
   */
  getEventTypeNames(): Record<WebhookEventType, string> {
    return {
      [WebhookEventType.USER_CREATED]: 'User Created',
      [WebhookEventType.USER_UPDATED]: 'User Updated',
      [WebhookEventType.USER_DELETED]: 'User Deleted',
      [WebhookEventType.USER_SIGNED_IN]: 'User Signed In',
      [WebhookEventType.USER_SIGNED_OUT]: 'User Signed Out',
      [WebhookEventType.SESSION_CREATED]: 'Session Created',
      [WebhookEventType.SESSION_EXPIRED]: 'Session Expired',
      [WebhookEventType.ORGANIZATION_CREATED]: 'Organization Created',
      [WebhookEventType.ORGANIZATION_UPDATED]: 'Organization Updated',
      [WebhookEventType.ORGANIZATION_DELETED]: 'Organization Deleted',
      [WebhookEventType.ORGANIZATION_MEMBER_ADDED]: 'Organization Member Added',
      [WebhookEventType.ORGANIZATION_MEMBER_REMOVED]: 'Organization Member Removed'
    };
  }

  /**
   * Calculate delivery success rate
   */
  calculateSuccessRate(stats: {
    total_deliveries: number;
    successful: number;
    failed: number;
  }): {
    rate: number;
    status: 'excellent' | 'good' | 'fair' | 'poor';
  } {
    if (stats.total_deliveries === 0) {
      return { rate: 0, status: 'poor' };
    }

    const rate = (stats.successful / stats.total_deliveries) * 100;

    let status: 'excellent' | 'good' | 'fair' | 'poor';
    if (rate >= 95) status = 'excellent';
    else if (rate >= 90) status = 'good';
    else if (rate >= 75) status = 'fair';
    else status = 'poor';

    return { rate, status };
  }

  /**
   * Format webhook endpoint for display
   */
  formatEndpoint(endpoint: WebhookEndpoint): {
    displayUrl: string;
    eventCount: number;
    statusText: string;
    isHealthy: boolean;
  } {
    // Truncate URL for display
    let displayUrl = endpoint.url;
    if (displayUrl.length > 50) {
      displayUrl = displayUrl.substring(0, 47) + '...';
    }

    return {
      displayUrl,
      eventCount: endpoint.events.length,
      statusText: endpoint.is_active ? 'Active' : 'Inactive',
      isHealthy: endpoint.is_active
    };
  }
}